examples/fluids/src/setuplibceed.c

```diff
@@ -213,33 +210,230 @@ PetscErrorCode SetupBCQFunctions(Ceed ceed, PetscInt dim_sur, PetscInt num_comp_
+  }
+
+  // Get cell interp, grad, and quadrature data
+  PetscCall(DMGetFirstLabeledPoint(dm, dm, domain_label ? domain_label : depth_label, 1, ids, height, &first_point, NULL));
+  PetscCall(DMPlexGetCellType(dm, first_point, &cell_type));
+  elem_topo = ElemTopologyP2C(cell_type);
+  if (!elem_topo) SETERRQ(PetscObjectComm((PetscObject)dm), PETSC_ERR_SUP, "DMPlex topology not supported");
+  {
+    size_t             q_points_size;
+    const PetscScalar *q_points_petsc;
+    PetscInt           q_dim;
+
+    PetscCall(PetscQuadratureGetData(quadrature, &q_dim, NULL, &Q, &q_points_petsc, &q_weights));
+    q_points_size = Q * dim * sizeof(CeedScalar);
+    PetscCall(PetscCalloc(q_points_size, &q_points));
+    for (PetscInt q = 0; q < Q; q++) {
+      for (PetscInt d = 0; d < q_dim; d++) q_points[q * dim + d] = q_points_petsc[q * q_dim + d];
+    }
+  }
+
+  {  // Convert to libCEED orientation
+    PetscBool       is_simplex  = PETSC_FALSE;
+    IS              permutation = NULL;
+    const PetscInt *permutation_indices;
+
+    PetscCall(DMPlexIsSimplex(dm, &is_simplex));
+    if (!is_simplex) {
+      PetscSection section;
+
+      // -- Get permutation
+      PetscCall(DMGetLocalSection(dm, &section));
+      PetscCall(PetscSectionGetClosurePermutation(section, (PetscObject)dm, dim, num_comp * P, &permutation));
+      PetscCall(ISGetIndices(permutation, &permutation_indices));
+    }
+
+    // -- Copy interp, grad matrices
+    PetscCall(PetscCalloc(P * Q * sizeof(CeedScalar), &interp));
+    PetscCall(PetscCalloc(P * Q * dim * sizeof(CeedScalar), &grad));
+    const CeedInt c = 0;
+    for (CeedInt q = 0; q < Q; q++) {
+      for (CeedInt p_ceed = 0; p_ceed < P; p_ceed++) {
+        CeedInt p_petsc = is_simplex ? (p_ceed * num_comp) : permutation_indices[p_ceed * num_comp];
+
+        interp[q * P + p_ceed] = basis_tabulation->T[0][((face * Q + q) * P * num_comp + p_petsc) * num_comp + c];
+        for (CeedInt d = 0; d < dim; d++) {
+          grad[(d * Q + q) * P + p_ceed] = basis_tabulation->T[1][(((face * Q + q) * P * num_comp + p_petsc) * num_comp + c) * dim + d];
+        }
+      }
+    }
+
+    // -- Cleanup
+    if (permutation) PetscCall(ISRestoreIndices(permutation, &permutation_indices));
+    PetscCall(ISDestroy(&permutation));
+  }
+
+  // Finally, create libCEED basis
+  CeedBasisCreateH1(ceed, elem_topo, num_comp, P, Q, interp, grad, q_points, q_weights, basis);
+  PetscCall(PetscFree(q_points));
+  PetscCall(PetscFree(interp));
+  PetscCall(PetscFree(grad));
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+// -----------------------------------------------------------------------------
+// Get CEED Basis from DMPlex
+// -----------------------------------------------------------------------------
+PetscErrorCode CreateBasisFromPlex(Ceed ceed, DM dm, DMLabel domain_label, CeedInt label_value, CeedInt height, CeedInt dm_field,
+                                   CeedQuadMode quad_mode, CeedBasis *basis) {
```

Comment:
This `CeedQuadMode` input won't work as you want when you start using simplex meshes. I don't think I would add this argument, as its saying to explicitly ignore the quadrature space of the FE and use something else

Reply:
Yeah, I thought it was used somewhere for `CEED_GAUSS_LOBATTO`, but forgot that all that went away with `CeedBasisCreateProjection`.

Reply:
Fixed.